Short-Answer/Algorithms_Answers.md

```diff
@@ -2,14 +2,42 @@
 
 ## Exercise I
 
-a)
+```
+a)  a = 0                       # O(1)
+    while (a < n * n * n):      # O(n)
+      a = a + n * n             # O(n)
+```
+a) There is one loop (the a < n), run up to n times, giving us an O(n).
 
 
-b)
+```
+b)  sum = 0                     # O(1)
+    for i in range(n):          # O(n)
+      j = 1                     # O(1)
+      while j < n:              # O(n)
+        j *= 2                  # O(1)
+        sum += 1                # O(1)
+```
+b) There are two for loops that are run n times, one is nested inside the other, so the 'n's are multiplied, giving is an O(n**2)
```

Comment:
Nice job with these.

This one is close, but for b), while there _is_ a loop within a loop you have to consider what happens to `j` on each iteration of the inner while-loop:

```
b)  sum = 0
    for i in range(n):
      j = 1
      while j < n:
        j *= 2
        sum += 1
```

`j` doubles in value on each iteration for `j < n`, therefore we half the number of elements we traverse through on each iteration of the inner while-loop. This would produce a runtime complexity of `O(log n)`

If you account for the outer for-loop, you get `O(n * log n)`